zkevm-circuits/src/blake2b/circuit/src/lib.rs

```diff
@@ -0,0 +1,1855 @@
+            g(&mut v, 2, 7, 8, 13, m[s[12]], m[s[13]]);
+            g(&mut v, 3, 4, 9, 14, m[s[14]], m[s[15]]);
+        }
+    
+        for i in 0..8 {
+            h[i] ^= v[i] ^ v[i + 8];
+        }
+    
+        h
+    }
+}
+
+// A small module providing functions for performing operations in the group of permutations of
+// the set {0, 1, ..., L - 1} for the specified L. A permutation, which a module function takes  
+// or returns, is represented by the array, the i-th entry of which contains the value replacing i  
+pub mod permutation {
+    // Computes the permutation composition a * b, i.e. the permutation, which is 
+    // equivalent to applying the permutations b and a in the corresponding order  
+    pub fn compose<const L: usize>(a: &[usize; L], b: &[usize; L]) -> [usize; L] {
+        let mut result = [0; L];
+        for i in 0..L {
+            result[i] = a[b[i]];
+        }
+        result
+    }
+
+    // Computes the inverse permutation for x
+    pub fn invert<const L: usize>(x: &[usize; L]) -> [usize; L] {
+        let mut result = [0; L];
+        for i in 0..L {
+            result[x[i]] = i;
+        }
+        result
+    }
+}
+
+// Converts a specified value into an Expression instance
+fn gf<F:FieldExt>(value: u128) -> Expression<F> {
+    Expression::Constant(F::from_u128(value))
+}
+
+// Converts a specified value into a Value instance
+fn known<F:FieldExt>(value: u64) -> Value<F> {
+    Value::known(F::from(value))
+}
+
+// Computes the index of a row, using it's offset relative to the specified row 
+fn shift_row(row: usize, offset: i32) -> usize {
+    let row = row as i64 + offset as i64;
+    assert!(row >= 0, "The row {} does not exist. Row indices are nonnegative!", row);
+    row as usize
+}
+
+// Computes the expression equal to the number stored in the specified Chunk 
+fn chunk_to_number<F:FieldExt, const H:u8>(meta: &mut VirtualCells<F>, chunk: Chunk<F,H>) -> Expression<F> {
+    (0..(H as usize)).map(|cell| chunk.expr(meta, cell)).rfold(gf(0), |sum, byte| { sum * gf(256) + byte })
+}
+
+// Splits the specified slice into arrays without copying its elements 
+fn arrefs_from_slice<T, const S: usize, const L: usize>(slice: &[T]) -> [&[T; S]; L] {
+    array::from_fn(|i| slice[S * i..][..S].try_into().unwrap())
+}
+
+// Creates Chunks, the byte columns for which are specified by "pairs". The relative places of these Chunks can be considered as cells 
+// of a table-like structure, for which columns are described by "pairs" and each row has a height of H cells. The cells of this structure
+// are numbered from left to right row by row. The relative position of the first cell of the structure is the row "offset" of the column 
+// "pairs[0][0]". The created chunks occupy the cells of a table-like structure, whose numbers are "skip", "skip" + 1, ..., "skip" + L - 1   
+fn create_chuncks<F:FieldExt, const H:u8, const L:usize>(pairs: &[[Column<Advice>; 2]], offset: i32, skip: usize) -> [Chunk<F,H>; L] {
```

Comment:
nit: chunk

Reply:
What do you mean?

Reply:
its a typo, probably should be chunk

Reply:
No. It returns [Chunk<F,H>; L], thus we have a plural form in the description.